### PR TITLE
Add FP8 support for batch_prefill with per-tensor quantization

### PR DESCRIPTION
**Pull request by poyenc**
## Summary
This PR adds FP8 (float8_e4m3) support to the `batch_prefill` kernel with per-tensor quantization, enabling efficient mixed-precision attention for paged KV cache scenarios.

## Changes
- **Core Implementation**:
  - Extended `mha_batch_prefill` to support FP8 input with BF16 output (`fp8bf16` mode)
  - Added per-tensor quantization support via `q_descale`, `k_descale`, `v_descale` parameters
  - Updated kernel selection logic to handle FP8 data type and quantization scale types
  
- **API Changes**:
  - Added optional `q_descale`, `k_descale`, `v_descale` parameters to `mha_batch_prefill_func`
  - Updated C++ interface to support `quant_scale_enum` (per-tensor vs no-scale)
  - Extended Python bindings with descale parameters

- **Testing**:
  - Added comprehensive FP8 test suite (`test_batch_prefill_fp8_output`) comparing FP8 vs BF16 kernels
  - Added cross-validation test comparing batch_prefill (paged) vs flash_attn_varlen implementations
  - Verified correctness across multiple configurations (causal/non-causal, various sequence lengths, logits
 
## Test Plan

All existing batch_prefill tests pass
- ✅ New FP8 tests validate output accuracy within threshold (max_diff < 0.055)
- ✅ Cross-validation between batch_prefill and flash_attn_varlen implementations
- ✅ Tested with various configurations: batch sizes, sequence lengths, causal modes, logits soft cap

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

**Comment by poyenc** [comment]
Need to move the batch_prefill fp8 tests in _test_batch_prefill_vs_varlen.py_ to _test_batch_prefill.py_

File: op_tests/test_batch_prefill_vs_varlen.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+"""
+Test to compare batch_prefill (paged KV cache) with flash_attn_varlen (varlen format).
+Both should produce identical results since they use the same qr_async pipeline.
+"""
+
+import torch
+import pytest
+import aiter
+from aiter import dtypes
+from aiter import per_tensor_quant
```
[the remark this answers is not in the document]
Comment (poyenc) [reply op_tests/test_batch_prefill_vs_varlen.py]: this is not a typo

File: op_tests/test_batch_prefill.py [hunk 2]
```diff
@@ -270,6 +273,263 @@ def convert_lens_to_indtpr(lens):
...
+                kv_page_indices,
+                max_seqlen_q,
+                max_seqlen_k,
+                causal=causal,
+                logits_soft_cap=logits_soft_cap,
+                q_descale=q_descale,
+                k_descale=k_descale,
+                v_descale=v_descale,
+            ),
+            0.1,
```
[the remark this answers is not in the document]
Comment (poyenc) [reply op_tests/test_batch_prefill.py]: will remove this

File: csrc/py_itfs_ck/mha_batch_prefill_kernels.cu [hunk 3]
```diff
@@ -175,18 +181,42 @@ mha_batch_prefill(at::Tensor& q,                  // [total_q, hq, d]
...
 
     TORCH_CHECK(k.dtype() == q_dtype, "query and key must have the same dtype");
     TORCH_CHECK(v.dtype() == q_dtype, "query and value must have the same dtype");
     TORCH_CHECK(cu_seqlens_q.dtype() == torch::kInt32, "cu_seqlens_q must have dtype int32");
     TORCH_CHECK(kv_indptr.dtype() == torch::kInt32, "kv_indptr must have dtype int32");
 
-    std::string q_dtype_str = q_dtype == torch::kFloat16 ? "fp16" : "bf16";
+    std::string dtype_str;
+    if(q_dtype == at::ScalarType::Half)
+        dtype_str = "fp16";
```
Comment (valarLip) [inline csrc/py_itfs_ck/mha_batch_prefill_kernels.cu]: this one https://github.com/ROCm/aiter/blob/4ca577a88fe394ca8353ef71897dd695d15ad18b/csrc/include/py_itfs_common.h#L65C20-L65C35 can be used for convert torch_dtype to str

File: csrc/py_itfs_ck/mha_batch_prefill_kernels.cu [hunk 4]
```diff
@@ -281,18 +311,21 @@ mha_batch_prefill(at::Tensor& q,                  // [total_q, hq, d]
     CHECK_SHAPE(kv_indptr, batch_size + 1);
     auto opts = q.options();
 
+    auto out_type = dtype_str == "fp8bf16" ? at::ScalarType::BFloat16 : q_dtype;
```
Comment (valarLip) [inline csrc/py_itfs_ck/mha_batch_prefill_kernels.cu:311]: so we only have bf16 supported for fp8 kvcache?

Comment (poyenc) [reply csrc/py_itfs_ck/mha_batch_prefill_kernels.cu:311]: Yes, we can add fp16 support if we need. Since the flow is identical except the output conversion part.

File: csrc/py_itfs_ck/mha_batch_prefill_kernels.cu (hunk 3 above)
Comment (poyenc) [reply csrc/py_itfs_ck/mha_batch_prefill_kernels.cu]: got it! I'll rewrite this part

File: csrc/py_itfs_ck/mha_batch_prefill_kernels.cu (hunk 4 above)
Comment (poyenc) [reply csrc/py_itfs_ck/mha_batch_prefill_kernels.cu:311]: As discussed, fp8 input / fp16 output support will be handled in a separate PR. This PR only includes bf16 support.

**Review by valarLip — APPROVED** [review]
